Migrate VPA metrics to CustomResourceState metrics and upgrade `kube-state-metrics` to `v2.13.0`

vicwicker opened the pull request:
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:

This PR allows the upgrade of the kube-state-metrics to the latest version, as of today v2.13.0. The upgrade was not possible because of the deprecation of the VPA metrics, which are now supposed to be configured as any other custom resource by using a so-called [`CustomResourceState`](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md) feature from the kube-state-metrics. This is a mechanism to tell kube-state-metrics via configuration which keys and values from custom resource spec files need to be mapped into which metrics. 

We tried to preserve the old metric labels and naming as much as possible. However, there are few changes. The most significant is that the resource (i.e., CPU or memory) is not a label anymore in the new metrics but part of the metric name.

<details><summary>Example for the blackbox-exporter CPU target recommendation</summary>

```
{
    "metric": {
      "__name__": "kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target",
      "container": "blackbox-exporter",
      "instance": "kube-state-metrics",
      "job": "kube-state-metrics",
      "namespace": "kube-system",
      "resource": "cpu",
      "target_api_version": "apps/v1",
      "target_kind": "Deployment",
      "target_name": "blackbox-exporter",
      "type": "shoot",
      "unit": "core",
      "verticalpodautoscaler": "blackbox-exporter"
    }
},
```

becomes

```
{
  "__name__": "kube_customresource_verticalpodautoscaler_status_recommendation_containerrecommendations_target_cpu",
  "container": "blackbox-exporter",
  "customresource_group": "autoscaling.k8s.io",
  "customresource_kind": "VerticalPodAutoscaler",
  "customresource_version": "v1",
  "instance": "kube-state-metrics",
  "job": "kube-state-metrics",
  "namespace": "kube-system",
  "target_api_version": "apps/v1",
  "target_kind": "Deployment",
  "target_name": "blackbox-exporter",
  "type": "shoot",
  "unit": "core",
  "verticalpodautoscaler": "blackbox-exporter"
}
```

</details>

This PR also improves the deployment of kube-state-metrics. In short, the most relevant enhancements are:

1. Move away from the concept of `ResourceConfigs` and adopt `ManagedResource` to deploy all kube-state-metrics artifacts. Code-wise this resembles now the way Prometheus is deployed.
2. Use the kube-state-metrics filtering capabilities to make the Prometheus scrape jobs lighter. Currently, unwanted metrics are also scraped but discarded during the scrape job relabeling step.

As for the unit tests, the `CustomResourceState` expected configuration is stored in a YAML file and committed to the repository. The `expectedCustomResourceStateConfig()` helper function is to be used when loading such expectation from disk during test execution. It loads this file as well as it generates the configuration the code being tested produces. This actual configuration is stored in a file in the temporary directory and matched against the expectation. If the match succeeds, the test will run. If not, the test will be aborted and a help message will be printed by ginkgo with a `diff` command to compare both files. Developers can compare both files with any tool of their choice and fix the test accordingly. Reviewers can find below and example of this test output.

<details>
<summary>Example of test failure</summary>

![Screenshot 2024-06-06 at 16 30 03](https://github.com/gardener/gardener/assets/4339456/3dd4e2fe-9975-41ca-998b-44922c991934)

</details>

**Special notes for your reviewer**:

/cc @istvanballok @rickardsjp 
/fyi @rfranzke @voelzmo 

**Release note**:

```other operator
Migrate VPA metrics to CustomResourceState metrics and upgrade `kube-state-metrics` to `v2.13.0`.
```

vicwicker committed:
Upgrade KSM to latest version

vicwicker committed:
Adapt VPA metrics to new solution CustomResourceState

This commit adds the necessary configuration changes to readd the VPA
metrics using the new strategy CustomResourceState from the
kube-state-metrics. Such configuration is created programmatically
thanks to the the kube-state-metrics dependency. This commit also runs
`go mod tidy` in order to make this dependency direct.

The new CustomResourceState configuration is stored in a ConfigMap and
mounted into the kube-state-metrics deployment using the
`--custom-resource-state-config-file`. There is an alternative command
flag `--custom-resource-state-config` to pass such configuration inline
but the resulting configuration is very lengthy, which would make the
kube-state-metrics deployment spec file not so readable.

We tried to preserve the old metric labels and naming as much as
possible. However, there are a few changes:

1. We can't have one common metric for several resources (e.g., CPU and
   memory) and distinguish between them via labels. Instead, we need to
   migrate to one specific metric per resource. For instance, previously
   we had a target metric with labels `resource=cpu` or
   `resource=memory` to indicate the value corresponding to CPU or
   memory. After this change two metrics are added: target_cpu and
   target_memory. Nevertheless, we consider this as an improvement
   because values in the new metrics are not mixing units as before
   (cores were used for the CPU and bytes for the memory).
2. Using CustomResourceState prefixes the metric name with
   kube_customresource by default. Therefore, the old prefix
   kube_verticalpodautoscaler becomes now
   kube_customresource_verticalpodautoscaler. Such default prefix can
   be overwritten to match the one we want, but we choose not to so that
   the new name reflects better that the metric is generated by the
   CustomResourceState.
3. The property `nilIsZero` is set to true for the recommendation
   metrics to set the value to zero if the recommendation path does not
   exist in the VPA spec file (i.e., it doesn't have a recommendation).
   In the past, the time series would simply not exist. Since a
   recommendation to 0 does not make sense, this is a way, e.g., in the
   dashboard, to know there is no recommendation at all, rather than
   relying on not having data, which could also be the case if there is
   an issue and metrics are missing.
4. Metrics containing "container_policies" in their name have been
   renamed to "containerpolicies" to respect the containerPolicies key
   in the spec file (similar to resourcePolicy or
   containerRecommendations). This way, the metric format is unified to
   use underscores to separate keys in the path.
5. The CustomResourceState documentation does not mention that
   kube-state-metrics needs `get` permissions on RBAC-enabled clusters,
   but only `list` and `watch`. Therefore, the old `get` permissions are
   removed.

As for unit testing, the expected CustomResourceState config is saved
into a file and committed to git. The
`expectedCustomResourceStateConfig()` generates a new config and loads
such file. This function is to be used to load the expected config
during the test setup. If the newly generated config matches the
expectation, then the test execution continues. Otherwise, the test
execution aborts. The generated config is saved into a file in a
temporary directory and a custom error message is shown with a hint on
how to use the diff command to see the diff between the actual and
expected configs. This of course only works on UNIX systems. Developers
can review this diff and react accordingly e.g., check if there is a bug
in the code or if indeed the expectation has to be changed.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Adapt VPA dashboard to new metrics

Queries will look lengthy for a short period of time because new metrics
are queried using the or expression so that visualizations do not break.
Once Prometheus instances have dropped the old metrics, we can simplify
these dashboard queries by removing the previous ones.

The templating can always use `_cpu` metrics because the new VPA
metrics have `nilIsZero` set to true. That means the time series always
exists for all VPAs, whether they have a target CPU or not. Note we could
take any other metric as an example.

vicwicker committed:
Add dependency k8s.io/kube-state-metrics/v2 v2.11.0

We can't upgrade to the latest k8s.io/kube-state-metrics/v2 due to a
version conflict with k8s.io/component-base v0.29.5.

The next version of k8s.io/kube-state-metrics/v2 requires a version of
github.com/prometheus/common (v0.51.1) that contains a breaking change.
This breaking change is incompatible with the version of
k8s.io/component-base (v0.29.5) that we're using.

By using k8s.io/kube-state-metrics/v2 v2.11.0 instead, we can avoid this
issue, as this version does not require the breaking change in
github.com/prometheus/common.

This change should not affect the runtime behavior, as the changes
between k8s.io/kube-state-metrics/v2 v2.11.0 and higher do not affect
the YAML serialization of the custom resource state data structures.

Once a new version of kubernetes/component-base is released containing
the fix
https://github.com/kubernetes/component-base/commit/62c04b35eff68f7d4087f94b7b1228facb375141
for the breaking change
https://github.com/prometheus/common/commit/bd41eb6b9dee4fa983f31ae8756700efde1f3ea2
that was introduced in v0.48.0 of prometheus/common, we can upgrade the
kube-state-metrics/v2 beyond v2.11.0.

Otherwise when using kube-state-metrics/v2 in the latest version the
build fails with a compilation error.

```
go get k8s.io/kube-state-metrics/v2@v2.11.0
```

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

rickardsjp commented on pkg/component/observability/monitoring/kubestatemetrics/resources.go:
```diff
@@ -37,6 +37,156 @@ import (
+showed the kind is superfluous when --custom-resource-state-config is given. We need a bit more investigation
+in case there is a resource of kind CustomResourceStateMetrics that we could use to pass this lenghty
+configuration more elegantly.
+*/
```
Have you had a look [at the API](https://pkg.go.dev/k8s.io/kube-state-metrics/v2@v2.12.0/pkg/customresourcestate)? Looks like it should be possible to avoid string literals and do it in code :)

vicwicker replied:
Yes. I moved to generate this configuration programatically some commits later :)

rickardsjp replied:
Oops, that's what I get for going through the commits but stopping short of the end.

vicwicker committed:
Extract dependencies with commented versions to their own require block

The next commits will use the new dependency kube-state-metrics, and
`go mod tidy` will convert this dependency into a direct one.

Because of the controller-manager dependency with a commented version,
`go mod tidy` will identify the new kube-state-metrics dependency
belongs to none of the existing require blocks and it will attempt to
create a new block just for kube-state-metrics. Note that the default
behaviour of go.mod is to keep two blocks: one for direct dependencies
and one for indirect.

This commit extracts this third type of dependency (dependency with
commented version) into its own require block. This way, running
`go mod tidy` in the following commits will preserve the three blocks
defined.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Change namespace used in metric for control-plane shoot pods

The kube-system namespace was previously used. This was wrong in two
ways. First, those pods are not deployed in the kube-system of the
shoot, but in a specific namespace in the seed. Second, there is a
blackbox exporter in both the shoot's control-plane namespace in the
seed and in the actual kube-system namespace in the shoot. This led to
naming conflicts and possibly unexpected query results, e.g. in the VPA
dashboard when checking the VPA recommendations for the blackbox
exporter pods. The namespace is changed to shoot-control-plane, which
properly describes what it is.

Note that the control plane namespace is technically called
`shoot--<project>--<shoot name>`, but instead of leaking this
implementation detail to the monitoring stack, we use the more
user-friendly name `shoot-control-plane`.

vicwicker committed:
Fix VPA recommendations dashboards for the Prometheus custom resource

Since the introduction of the Prometheus operator as part of GEP-19, the
VPA targets the Prometheus custom resource which is called `shoot`. The
Prometheus pods are prefixed with "prometheus-", e.g.
`prometheus-shoot-0`, which breaks the VPA dashboard because the
assumption that the VPA target name is a prefix of the pod name is no
longer true: in this case the VPA target kind (`Prometheus`) is added as
a prefix to the pod name in lower case. So we use the `?i` flag to
expect the target kind to be an optional prefix of the pod name.

vicwicker committed:
Re-introduce the kube_pod_labels metric

The upgrade to higher kube-state-metrics versions disable the
kube_pod_labels metric by default, which now needs to be re-enabled
explicitly. Only the origin label is required by Gardener.

This metric is used in the following two alerting rules:

- alert: KubePodPendingShoot
  expr: (kube_pod_status_phase{phase="Pending", type="shoot"} == 1 and on(pod) kube_pod_labels{label_origin="gardener"})
- alert: KubePodNotReadyShoot
  expr: (kube_pod_status_ready{condition="true", type="shoot"} == 0 and on(pod) kube_pod_labels{label_origin="gardener"})

This was broken since the upgrade to kube-state-metrics 2.5.0 where the
kube_pod_labels metric existed but did not have an origin label unless
explicitly configured.

vicwicker committed:
Allow metric for the VPA uncapped target

Although this metric was available in previous versions of
kube-state-metrics, it was never allowed. However, we have identified it
as relevant for certain use-cases (e.g., we plan to use it for follow-up
alerts) and, hence, we add it here to the allow list.

vicwicker committed:
KSM server-side filter for allowed metrics

kube-state-metrics can be configured to expose only specific metrics
about the state of objects. In Gardener, we are not interested in
ingesting all available metrics into Prometheus but a subset of them.
Previously, this filtering was achieved via the scrape configuration on
the Prometheus side.

The kube-state-metrics in the seed is already configured to expose
certain metrics, but the kube-state-metrics in the shoot is not. Thus,
when the shoot Prometheus scrapes the kube-state-metrics exposed
metrics, it downloads a bunch of unwanted metrics and filters them out
via metric relabeling. This is inefficient because such filtering can be
done server-side to reduce the scrape time.

This commit improves this by extending the kube-state-metrics
configurations to make sure server-side filtering is in-place. This is
done via the `--metric-allowlist` flag, which expects a comma-separated
list of regular expressions. Metric names that match at least one of
those regular expressions are exposed. Therefore, metric names in
`--metric-allowlist` are now wrapped around the ^ and $ characters to
fully match the metric name.

The client-side filtering from the Prometheus scrape configuration is
now removed as it is not necessary. The kube-state-metrics deployments
expose only the desired metrics and Prometheus injects everything that
is scraped. Keeping the client-side filtering could be seen as a second
line of defense, but it would also add a bit of configuration overhead
without a clear benefit.

As for the garden cluster, the kube-state-metrics is also deployed with
its specific allowed metric configuration. Previously, it was deployed
as for a regular seed and undesired metrics were filtered out by a
scrape config. VPA metrics for the garden cluster are also added to the
garden Prometheus as a preparation for further work to alert on certain
VPA recommendations, including resources in the garden namespace. In
consequence, the number of samples in the garden Prometheus increases.
There are around 250 containers deployed in the garden cluster and 13
VPA metrics per container. This adds up to 250*13=3250 new samples. This
is just a rough maximum estimation as not all containers are monitored
by VPA, so the real increase is probably somewhat lower. A side-effect
of ingesting metrics from the garden namespace is that the number of
samples for the kube_pod_container_status_restarts_total and
kube_pod_status_phase metrics also increases. However, this is also not
significant. Such metric cardinalities are five and one, respectively:
there are five pod states (pending, succeeded, failing, unknown and
running) and only one restart counter per pod. The garden cluster
deploys around 110 pods in the garden namespace. Consequently, this
increases the number of samples up to 110*(5+1)=660.

vicwicker committed:
Add a managed resource for the seed

This artifact will control the resources that are deployed in the
control-plane or the garden namespace in the seed.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Add a managed resource for the shoot

This artifact will control the resources that are deployed in the the
shoot. This is only needed for the control-plane kube-state-metrics.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the cluster role via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the service account via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the cluster role binding via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the service via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the deployment, VPA and PDB via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the cache, seed and garden scrape configs via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the shoot scrape config via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the prometheus rule via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Manage the custom resource state config via the managed resource

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Remove the resources configs

All resources are now managed by the managed resource.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Make the custom resource state config map immutable

So that when the config map changes, the deployment is rolled and the
config is not updated in place.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Wait for previous managed resource to be healthy

The cluster role and cluster role binding were originally managed by the
shoot-core-kube-state-metrics managed resource but are now managed by
the shoot-core-kube-state-metrics-target managed resource.

In order to migrate those two resources to the latter managed resource,
we need to wait until the former stops managing them and deletes them.
Otherwise, the previous managed resource wants to delete them while the
new one keeps creating them.

As for the tests, a second goroutine is introduced to watch to the
shoot-core-kube-state-metrics managed resource and set its generation
and status conditions to healthy.

vicwicker committed:
Adapt unit tests for new managed resources

vicwicker committed:
Fix deprecated comment

vicwicker committed:
Remove flag IsWorkerless from the KSM code

Only shoots can be workerless and the botanist component already takes
care of skipping the kube-state-metrics component if the shoot is
workerless.

Hint: This commit is better reviewed by hiding the whitespace change.

vicwicker committed:
Extract function to list the KSM resources for the seed

vicwicker committed:
Extract function to list the KSM resources for the shoot

vicwicker committed:
Register KSM common resources separately

This prepares the following commits for refactoring to improve code
readability.

vicwicker committed:
Extract function to list the KSM resources for the shoot target

vicwicker committed:
Simplify too many if-statements for type shoot

This also inlines the `genericTokenKubeconfigSecretName` variable.

vicwicker committed:
Fix nitpick format

rfranzke commented on go.mod:
```diff
@@ -74,17 +74,19 @@ require (
 	sigs.k8s.io/yaml v1.4.0
 )
 
+require sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20231015215740-bf15e44028f9 // v0.16.3
```
Why did you move this out of the above `require` block? Please move it back.

vicwicker replied:
I tried to capture the reasons in commit [`d281b7b`](https://github.com/gardener/gardener/pull/9941/commits/d281b7ba2f0a7efad3065c98d464b13126d3b92b). 

But in short, when this line is inside the `require` block from above, go is not able to understand that such block is for the direct dependencies and that new direct dependencies should be added there as well. So, when I added the KSM dependency via `go get` and `go mod tidy`,  go created a new `require` block only for the KSM. @istvanballok and I investigated this and found the culprit is this dependency because it has the version commented at the end (`... // v0.16.3`). After extracting this dependency in a new `require` block, `go` correctly added the new KSM dependency in the right block.

But this was just a proposal. We can keep it in one block altogether. I will address this 👍🏻

rfranzke replied:
Does it work with just one block if the comment is not at the end but in the line before?

vicwicker replied:
I just tried. Unfortunately it does not :(

rfranzke replied:
OK. I can live with both, either put it back, or keep it in the other block, but then maybe add a short comment why this is (to prevent we move this back up later)

vicwicker replied:
I put it back. I think it's going to be more confusing than helpful :)

rfranzke commented on pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go:
```diff
@@ -107,30 +156,77 @@ func (k *kubeStateMetrics) Deploy(ctx context.Context) error {
+			return err
+		}
+
+		// TODO(vicwicker): Remove after KSM upgrade to 2.13.
```
```suggestion
		// TODO(vicwicker): Remove after Gardener v1.104 got released.
```

rfranzke replied:
If we want to safely remove this code in 3 releases, I guess we should add some extra deploy step for `kube-state-metrics` after https://github.com/gardener/gardener/blob/1832cc548ca2cf8eb9bd2b2454728f217ec14f01/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L132-L135 in case `kube-state-metrics` already exists in the shoot namespace.
This way, we can be sure that this code has been called for all shoots.

vicwicker replied:
This should be addressed now.

rfranzke commented on pkg/component/observability/monitoring/kubestatemetrics/resources.go:
```diff
@@ -736,3 +712,14 @@ func (k *kubeStateMetrics) nameSuffix() string {
+
+func (k *kubeStateMetrics) customResourceStateConfigMap() *corev1.ConfigMap {
+	customResourceStateConfig, err := yaml.Marshal(NewCustomResourceStateConfig())
+	utilruntime.Must(err)
```
We don't want to panic, please return the error instead.

vicwicker replied:
Yes, you're right. I didn't return the error in the first place because the configuration that could error here is statically generated. But it definitely feels better to return the error and be on the safe-side 👍🏻

rfranzke commented on pkg/component/observability/monitoring/kubestatemetrics/resources.go:
```diff
@@ -736,3 +712,14 @@ func (k *kubeStateMetrics) nameSuffix() string {
+	cm := &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: customResourceStateConfigMapNamePrefix, Namespace: k.namespace}}
+	cm.Data = map[string]string{
+		customResourceStateConfigMountFile: string(customResourceStateConfig),
+	}
```
```suggestion
	configMap := &corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{Name: customResourceStateConfigMapNamePrefix, Namespace: k.namespace},
		Data: map[string]string{
			customResourceStateConfigMountFile: string(customResourceStateConfig),
		},
	}
```

rfranzke commented on pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go:
```diff
@@ -77,21 +76,71 @@ type Values struct {
+	clusterRole := k.clusterRole()
+	serviceAccount := k.serviceAccount()
+	customResourceStateConfigMap := k.customResourceStateConfigMap()
+	deployment := k.deployment(serviceAccount, "", nil, customResourceStateConfigMap.Name)
```
var () block

rfranzke commented on pkg/component/observability/monitoring/kubestatemetrics/kubestatemetrics.go:
```diff
@@ -77,21 +76,71 @@ type Values struct {
+
+func (k *kubeStateMetrics) getResourcesForShoot(genericTokenKubeconfigSecretName string, shootAccessSecret *gardenerutils.AccessSecret) []client.Object {
+	customResourceStateConfigMap := k.customResourceStateConfigMap()
+	deployment := k.deployment(nil, genericTokenKubeconfigSecretName, shootAccessSecret, customResourceStateConfigMap.Name)
```
var () block

vicwicker replied:
After addressing [this comment](https://github.com/gardener/gardener/pull/9941#discussion_r1713636431) the code now looks like the following. Unfortunately, I am not sure if a var block is still meaningful after the change. I would rather leave it like this if that's okay with you.

```
func (k *kubeStateMetrics) getResourcesForShoot(genericTokenKubeconfigSecretName string, shootAccessSecret *gardenerutils.AccessSecret) ([]client.Object, error) {
	customResourceStateConfigMap, err := k.customResourceStateConfigMap()
	if err != nil {
		return nil, err
	}

	deployment := k.deployment(nil, genericTokenKubeconfigSecretName, shootAccessSecret, customResourceStateConfigMap.Name)
	return []client.Object{
		...
	}, nil
}
```

rfranzke commented on pkg/component/observability/monitoring/kubestatemetrics/resources.go:
```diff
@@ -544,168 +582,104 @@ func (k *kubeStateMetrics) reconcileScrapeConfigGarden(scrapeConfig *monitoringv
+			Expr:   intstr.FromString(`sum(kube_node_info) by (os_image, kernel_version)`),
+		},
+		// Mitigation for extension dashboards.
+		// TODO(istvanballok): Remove in a future version. For more details, see https://github.com/gardener/gardener/pull/6224.
```
When can this be removed exactly?

vicwicker replied:
Unfortunately I can't answer that myself. Maybe we could ping @istvanballok and trigger the discussion.

rfranzke reviewed (changes requested):
Nice PR, thanks for the refactoring to GRM-only management :)

vicwicker committed:
Address review comments

- Move setup-envtest dependency back to the direct dependencies block.
- Return error instead of panicking when creating the
  CustomResourceState configuration.
- Move variable declarations into a var() block when possible.

Co-authored-by: Rafael Franzke <rafael.franzke@sap.com>

vicwicker committed:
Address review comments

Add additional KSM deploy step to ensure the migration runs for all
shoots.

Co-authored-by: Rafael Franzke <rafael.franzke@sap.com>

rfranzke commented on pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go:
```diff
@@ -133,6 +133,13 @@ func (r *Reconciler) runReconcileShootFlow(ctx context.Context, o *operation.Ope
+		// TODO(vicwicker): Remove after Gardener v1.104 got released.
+		_ = g.Add(flow.Task{
+			Name:         "Reconciling kube-state-metrics for Shoot in Seed for the monitoring stack during migration to Gardener v1.104",
+			Fn:           flow.TaskFn(botanist.DeployKubeStateMetrics).RetryUntilTimeout(defaultInterval, 2*time.Minute),
```
```suggestion
			Fn:           flow.TaskFn(func(ctx context.Context) error {
				// check here if the `kube-state-metrics` deployment exists in the shoot namespace. if not, return nil
				return botanist.DeployKubeStateMetrics(ctx)
			}).RetryUntilTimeout(defaultInterval, 2*time.Minute),
```

We don't want to prematurely deploy KSM into the shoot namespace if it does not exist yet. We just want to migrate it early if it already exists. Please adapt :)

vicwicker replied:
It should be addressed now.

rfranzke reviewed (commented):
Thank you @vicwicker !

/lgtm
/approve

rickardsjp reviewed (approved):
/lgtm
